azerothcore/azerothcore-wotlk
# feat(Core/Creatures): implement a sparring system

## Changes Proposed:

This PR proposes changes to:

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/15849

## SOURCE:

The changes have been validated through:

## Tests Performed:

This PR has been:


## How to Test the Changes:



The test needs some DB tweaks:
1. You need two creatures hostile to each other (for example, a stormwind guard against a wolf).
2. Add to creature_sparring the guid of creatures that will not receive damage below a certain percentage.
3. Look if it's working and check various situations (should receive damages if the player attacks or mindcontrols the victim or the attackers, for example).

DB Query (go to goldshire to test the change):
```
INSERT INTO `creature_sparring` (`GUID`, `SparringPCT`) VALUES
(4356, 75);

UPDATE `creature` SET `position_x` = -9481.81, `position_y` = 72.47, `position_z` = 56.4849 WHERE `guid` = 4356
``` 

## Known Issues and TODO List:


- [ ] I didn't do research on live servers, so without additional information, the behaviour is maybe not 100 % blizzlike.


## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
